### PR TITLE
PKGBUILD and DKMS to make it easier for arch users..

### PR DESCRIPTION
**Pull request by pxlsec**
I took the time to rewrite some of the build system to make it easier to install, and maybe even possible to publish to the AUR. Anyways, I did some changes to the Makefile inside of the driver folder, so I have no clue if the original install script works properly at the moment. But ill take a look when I get some time left over.

Cheers <3

### CODE REVIEW

**Comment by Gnarus-G** [comment]
@pxlsec Very much appreciate you working on this.

Now, I hate to be that guy but I want some better commit messages. Something like:
`Create a PKGBUILD to build packages`
The criteria roughly being:
- First letter is capitalized
- Imperative tense (e.g. "Add ...", not "Added ..." or "Adds ...")
- Include as much detail as you can.

And, lol, I love that you're using the one true editor but that doesn't need to be part of the commit message.

So please rebase those commits when you can.

**Comment by pxlsec** [comment]
> Now, I hate to be that guy but I want some better commit messages.
Haha, don't worry. As long as you had a good laugh I'm happy. Besides I need to learn how to use git & github properly anyways.
I also continued working on it a little bit and it feels decent now..

The good:
- I can install the driver on my custom kernel build using clang.
- All files are managed by the package, aka uninstalls are completely clean.
- DKMS so that you don't have to manually install after every update.
- Also installs to all kernels when installing the package.

The bad
- The PKGBUILD currently pulls the whole repo again, even tho it's in the repo. This can be fixed quite easily but I don't know whether you want the package to be on the AUR or not.
- DKMS only works when using the PKGBUILD (for now).
- The module does not install itself, so you have to run `modprobe maccel` after install. It is highly discouraged to automate according to the [arch wiki](https://wiki.archlinux.org/title/DKMS_package_guidelines#Module_loading_automatically_in_.install).
- Debug builds are currently broken?

**Comment by Gnarus-G** [comment]
Seems promising. Btw is it not possible to reuse the current install/uninstall scripts in the dkms configs?
And remember to rebase onto main because I merged #35 recently.

**Comment by MCPO-Spartan-117** [comment]
Since this will be managed by the package manager it would be fine to use `/usr/bin` for the binaries.

**Comment by MCPO-Spartan-117** [comment]
> The PKGBUILD currently pulls the whole repo again, even tho it's in the repo. This can be fixed quite easily but I don't know whether you want the package to be on the AUR or not.

You can just download the PKGBUILD directly.

**Comment by MCPO-Spartan-117** [comment]
> The criteria roughly being:
> 
>     * First letter is capitalized
> 
>     * Imperative tense (e.g. "Add ...", not "Added ..." or "Adds ...")
> 
>     * Include as much detail as you can.

Perhaps make a Contribution section in the readme or separate file for this.

**Comment by Gnarus-G** [comment]
BTW @MCPO-Spartan-117 and @pxlsec, you're welcome to join [maccel's discord server](https://discord.gg/zfNw39R4). I'd be happy to chat with you there should you ever want to.

File: PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    sed -e "s/@_PKGNAME@/${_pkgname}/" \
+        -e "s/@PKGVER@/${pkgver}/" \
+        -i "${pkgdir}/usr/src/${_pkgname}-${pkgver}/dkms.conf"
+
+    # Copy sources (including Makefile)
+    cp -r ${srcdir}/maccel/driver/* ${pkgdir}/usr/src/${_pkgname}-${pkgver}/
+
+    # Install CLI
+    install -Dm 755 $srcdir/maccel/cli/target/release/maccel ${pkgdir}/usr/local/bin/maccel
+    install -Dm 755 $srcdir/maccel/cli/usbmouse/target/release/maccel-driver-binder ${pkgdir}/usr/local/bin/maccel-driver-binder
```
Comment (gnusenpai) [inline PKGBUILD:39]: Files owned by packages shouldn't be installed into `/usr/local`. This directory is typically reserved for custom stuff by the system administrator.

File: driver/dkms.conf [hunk 2]
```diff
@@ -0,0 +1,7 @@
+PACKAGE_NAME="@_PKGNAME@"
+PACKAGE_VERSION="@PKGVER@"
+MAKE[0]="make KVER=$kernelver"
+CLEAN="make clean"
+BUILT_MODULE_NAME[0]="@_PKGNAME@"
+DEST_MODULE_LOCATION[0]="/kernel/drivers/usb"
```
[the remark this answers is not in the document]
Comment (pxlsec) [reply driver/dkms.conf:6]: Are you sure? I checked the source code and in every version i checked it installs to [/kernel/drivers/usb](https://github.com/Gnarus-G/maccel/blob/34a980690564e776a5cca9ba9f59ca11ee82c781/Makefile#L5)

File: maccel.install [hunk 3]
```diff
@@ -0,0 +1,19 @@
+post_install() {
+    udevadm control --reload-rules
+    udevadm trigger --subsystem-match=usb --subsystem-match=input --subsystem-match=hid --attr-match=bInterfaceClass=03 --attr-match=bInterfaceSubClass=01 --attr-match=bInterfaceProtocol=02
+
+    echo ' Add yourself to the "maccel" group\n'
+    echo ' usermod -aG maccel $USER\n'
+}
+
+post_uprgade() {
```
[the remark this answers is not in the document]
Comment (pxlsec) [reply maccel.install:9]: I have the fix on local, just not ready to commit quite yet. I was wondering if the module automatically reloads on an upgrade or if you need to `rmmod maccel && modprobe maccel`..

File: PKGBUILD (hunk 1 above)
Comment (pxlsec) [reply PKGBUILD]: Fixed in [d04fcc6](https://github.com/Gnarus-G/maccel/pull/36/commits/d04fcc66dc3135141a1fa2515029f7d6bb181355) which also introduced a typo (fixed on local)

**Comment by pxlsec** [comment]
I'm starting to feel quite happy with the PR. The only things I'm really missing is a better versioning system like `git describe` to automate version numbering, and a crash message for the cli ( currently it just segfaults if it can't access the kernel module). But I'm afraid that's outside of the scope of this project.

I'm opening the PR for reviews. For testing purposes, change out every instance of `github.com/Gnarus-G/maccel` to `github.com/pxlsec/maccel` since the main branch currently doesn't support dkms.

File: driver/dkms.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (MCPO-Spartan-117) [reply driver/dkms.conf]: I see where the confusion is then, the build system installs there but that isn't where the module originally was, where the module is though isn't really important, i'm just mentioning it.

File: install.sh [hunk 4]
```diff
@@ -93,12 +98,27 @@ install_driver() {
...
+    # Set name and version
+    sudo sed -e "s/@_PKGNAME@/maccel/" \
+         -e "s/@PKGVER@/${VERSION}/" \
+         -i "/usr/src/maccel-${VERSION}/dkms.conf"
+    
+    sudo cp -r $(pwd)/driver/* /usr/src/maccel-${VERSION}/
+
+    sudo dkms install "maccel/${VERSION}"
+
+    sudo modprobe maccel
```
Comment (MCPO-Spartan-117) [inline install.sh:114]: DKMS seems to already handle this on install in the `framework.conf` file.
```
       $modprobe_on_install
              Automatically load the built modules upon successful installation.
```

Comment (pxlsec) [reply install.sh:114]: Wow, I had no idea you could do that. I removed `modprobe` as suggested.
Is this setting something that should be mentioned in the README?

Comment (pxlsec) [reply install.sh]: I tested, It's not. I had to modprobe with it commented out. But it worked automatically with it enabled. 
I was also thinking that it would be a good improvement to mention adding the user to the maccel group in the README since the cli currently just crashes without it. And the only mention of it is in the install script (same for PKGBUILD).

**Comment by MCPO-Spartan-117** [comment]
Only things i can see this needs now is quoting variable paths and debug build handling for the CLI in the PKG, something like this.

```bash
if [ -d "$srcdir"/maccel/cli/target/release ]; then
    BUILDTAR="release"
elif [ -d "$srcdir"/maccel/cli/target/debug ]; then
    BUILDTAR="debug"
else
    echo "There's either no build or it's a different build than we are expecting."
    return 1
fi

    install -Dm 755 "$srcdir"/maccel/cli/target/"$BUILDTAR"/maccel "${pkgdir}"/usr/bin/maccel
    install -Dm 755 "$srcdir"/maccel/cli/usbmouse/target/"$BUILDTAR"/maccel-driver-binder "${pkgdir}"/usr/bin/maccel-driver-binder
```

Other than that it seems ready to merge.

**Comment by pxlsec** [comment]
Correct me if im wrong, but this way of handling debug builds seems like a very hacky approach and I'm not sure I like it..
- I don't think you should be able to build a package in multiple ways, it would destroy reproducability.
- The build mode specified in the Makefile is release, so you will not be able to install a debug build without editing the PKGBUILD anyways.
- The debug info will be stripped anyways depending on the users makepkg.conf (which I think is default behavior?)

However if I'm not mistaken, you should be able to build release mode with debug info. This way the PKGBUILD can strip to get a regular release and you could have debug info if enabled in makepkg.conf.

**Comment by MCPO-Spartan-117** [comment]
1. Then don't change the options or PKG at all?
2. Alright.
3. Less effort to deal with than 4
4. Is that something you can do without appending/removing 20 CFLAGs?

**Comment by pxlsec** [comment]
Yeah, It's done. I also added some stuff recommended by the [arch wiki](https://wiki.archlinux.org/title/Rust_package_guidelines#Prepare) for rust packages.

I tested building the package with `debug` and `strip` settings enabled, meaning that debug symbols will be stripped and put into a separate package ending with -debug, with the modes `Release`, `Debug` and `Release with Debug`.
| | Release | Debug | Release w/ Debug |
| - | ------------- | ------------- | ------------- |
| maccel-dkms| 1.5 MB | 2.5MB | 1.5 MB |
| maccel-dkms-debug| 178 KB | 7.7MB | 8.8 MB |

In summary, it seems to be working just fine. Further testing might be necessary however, as debugability might be worse with an optimized build.

**Review by MCPO-Spartan-117 — APPROVED** [review]
Haven't tested the PKG but i don't see any reason it would fail, DKMS works fine, the debug stuff and module path is up to @Gnarus-G.

File: uninstall.sh [hunk 5]
```diff
@@ -32,5 +38,6 @@ delete_everything() {
 }
 
 udev_uninstall
-delete_module
+#delete_module
```
Comment (Gnarus-G) [inline uninstall.sh:41]: So if we don't need this function anymore, I'll delete it.

File: README.md [hunk 6]
```diff
@@ -19,12 +19,18 @@ acceleration factor
...
-`curl`, `git`, `make`, `gcc`, and the linux headers in `/lib/modules/`
+`curl`, `git`, `make`, `gcc`, `dkms`, and the linux headers in `/lib/modules/`
 
 ```sh
 curl -fsSL https://www.maccel.org/install.sh | sudo sh
 ```
 
+Remember to `modprobe maccel` after installing or add `modprobe_on_install=true` to your dkms
+config file (usually located at /etc/dkms/framework.conf) to automatically modprobe after installing 
+a dkms module.
```
Comment (Gnarus-G) [inline README.md:30]: @pxlsec Is it bad to add `modprobe maccel` to the install script?

**Comment by Gnarus-G** [comment]
> Haven't tested the PKG but i don't see any reason it would fail, DKMS works fine, the debug stuff and module path is up to @Gnarus-G.

What module path are we talking about?

File: README.md (hunk 6 above)
Comment (Gnarus-G) [reply README.md:30]: I missed the points made earlier. I get this and I agree.

**Comment by Gnarus-G** [comment]
@pxlsec What's the whole point of the debug setting and where do you set it?
